Added Annotation Vectors Tutorial, #177

alvii147 opened the pull request:
Added first draft of the Annotation Vectors Tutorial, requested in issue #177, for feedback. The tutorial includes all 4 examples explored in the Matrix Profile V paper (https://www.cs.ucr.edu/~hdau001/guided_motif_search/). Planning to remove 2/3 out of the 4 examples. Let me know which ones I should keep/remove.

# Pull Request Checklist

Below is a simple checklist but please do not hesitate to ask for assistance!

- [x] Fork, clone, and checkout the newest version of the code
- [x] Create a new branch
- [x] Make necessary code changes
- [x] Install `black` (i.e., `python -m pip install black` or `conda install -c conda-forge black`)
- [x] Install `flake8` (i.e., `python -m pip install flake8` or `conda install -c conda-forge flake8`)
- [x] Install `pytest-cov` (i.e., `python -m pip install pytest-cov` or `conda install -c conda-forge pytest-cov`)
- [x] Run `black .` in the root stumpy directory
- [x] Run `flake8 .` in the root stumpy directory
- [x] Run `./setup.sh && ./test.sh` in the root stumpy directory
- [x] Reference a Github issue (and create one if one doesn't already exist)

seanlaw commented:
@alvii147 I did a quick pass over the tutorial and I really liked what you've done so far. Having said that, I stopped reading by the time I got to the EOG dataset as my attention was starting to drift (not because of the specific example itself). I will try to go through it again.

alvii147 commented:
Oh no I've bored you already! 😛
No worries, take your time, thanks for the feedback so far. I'm taking note of them.

seanlaw commented:
@alvii147 I was thinking about this last night and I am wondering if it makes sense to break this tutorial up into two parts (Part I and Part II)? Maybe the first part only focuses on the basics of annotation and the second part is more advanced?

alvii147 commented:
@seanlaw that makes sense actually, I remember considering this but was unsure if it was okay to write two tutorials for one paper. I can split this up to two tutorials for sure.

If I were to do that, then yes the first tutorial would have sort of an introduction to annotation vectors and guided motif searches in general. The second tutorial would then have more space for maybe 1-3 of the 4 examples. And because we have explored a lot of the theory and details in the first tutorial, the second one can be more on the results and less on the explanations.

Let me know what you think.

seanlaw commented:
> If I were to do that, then yes the first tutorial would have sort of an introduction to annotation vectors and guided motif searches in general. The second tutorial would then have more space for maybe 1-3 of the 4 examples. And because we have explored a lot of the theory and details in the first tutorial, the second one can be more on the results and less on the explanations.

Yes, I think that makes sense. Perhaps, instead of Part I and Part II, it might be "Annotation Vector Basics" and "Advanced Annotation Vectors"?

alvii147 commented:
Yepp that makes sense. I'm going to start separating the two, I'll keep you updated.

JaKasb commented:
I suggest the following edits:

- Remove the Dodgers Loop Dataset because it mostly contains pandas data wrangling.
  The produced step function wich servers as AV is never used.
-  If you want to keep it because the paper also contains the dataset, please shorten the pandas-wrangling:
```python
dodgers_df = pd.read_csv('https://archive.ics.uci.edu/ml/machine-learning-databases/event-detection/Dodgers.data', header=None)
dodgers_df.columns = ['DateTime', 'Count']
dodgers_df = dodgers_df[dodgers_df.Count != -1]

dodgers_df = dodgers_df.set_index(pd.to_datetime(dodgers_df['DateTime'], format="%m/%d/%Y %H:%M"))
dodgers_df.drop('DateTime', axis=1, inplace=True) # line is optional

annotation_vector = (dodgers_df.index.hour > 5) & (dodgers_df.index.hour < 12)
annotation_vector
```

- Move the fNIRS example to the top. This example is good because it shows how to construct the AV from data and how the compensated AV differs from the basic AV.
At cell [6], the plot contains only a visualization of the recordings. The same data visualization is also contained in the following plots, in cell [9] and in cell [10]. Therefore cell [6] is redundant and can be removed
- In the EOG example the text after cell 16 can be improved:
  "overflow and underflow points" -> clipping regions
  "variable used to store the data had the storage capacity of a signed 8-bit floating point number" ->
  "the measurement value exceeded the measurement range."
  (The dynamic range of the recordings is ± 124.51 . I assume that this is the limit of the ADC and not an `int8` limitation.)

alvii147 commented:
@JaKasb thanks for the feedback!

> Remove the Dodgers Loop Dataset because it mostly contains pandas data wrangling.
The produced step function wich servers as AV is never used.
If you want to keep it because the paper also contains the dataset, please shorten the pandas-wrangling:

So I included this part as an introduction to annotation vectors in general. As Sean mentioned above, we're gonna break this up into two parts, first part will be mostly basics of annotation vectors, and second part would be advanced usage.

As for the data wrangling, you're right it does look super clumsy. What we hopefully plan to do though, is upload the already-processed database to Zenodo once we're done. So the tutorial will actually just contain the read_csv function. The data will have already been processed.

> Move the fNIRS example to the top. This example is good because it shows how to construct the AV from data and how the compensated AV differs from the basic AV.
At cell [6], the plot contains only a visualization of the recordings. The same data visualization is also contained in the following plots, in cell [9] and in cell [10]. Therefore cell [6] is redundant and can be removed

Good catch, I guess my initial instinct was to visualize the data in general first, and then do stuff with it, but I think doing that makes it too lengthy, I'm gonna remove the first plot.

> In the EOG example the text after cell 16 can be improved:
"overflow and underflow points" -> clipping regions
"variable used to store the data had the storage capacity of a signed 8-bit floating point number" ->
"the measurement value exceeded the measurement range."
(The dynamic range of the recordings is ± 124.51 . I assume that this is the limit of the ADC and not an int8 limitation.)

Thanks for pointing these out, this sounds much better. I'm going to replace these phases.

Thanks again for your feedback! If you have any additional thoughts please feel free to shoot, I really appreciate it.

alvii147 commented:
### Split Annotation Vectors Tutorial to 2 parts

Annotation vectors tutorial was split into two parts, "Basics of Annotation Vectors" and "Advanced Annotation Vectors". Summary and Resources sections were added to both tutorials and the Advanced Annotation Vectors tutorial includes a quick recap section. Additionally, feedback from the previous pull request was used to improve the tutorial.

alvii147 commented:
@seanlaw I realized the reviewnb page does not show diffs, since I changed the filenames. So I manually generated diffs between the notebooks in the previous commit and the current commit, in case you want to look at only the diffs:

[notebook_diffs.zip](https://github.com/TDAmeritrade/stumpy/files/6678595/notebook_diffs.zip)

`diff_part1.html`:
diff between `Tutorial_Annotation_Vectors.ipynb` from previous commit & `Tutorial_Basics_Of_Annotation_Vectors.ipynb` from latest commit

`diff_part2.html`:
diff between `Tutorial_Annotation_Vectors.ipynb` from previous commit & `Tutorial_Advanced_Annotation_Vectors.ipynb` from latest commit

seanlaw commented:
Overall, I really like the "Basics" tutorial! Great work

seanlaw commented on docs/Tutorial_Advanced_Annotation_Vectors.ipynb:
```diff
@@ -0,0 +1,981 @@
+{
```
<p>Instead of "It summarizes...", I would just combine it with the previous sentence with `and it summarizes..."</p><p> </p><p>"and  replicates THE presented case studies"</p><p> </p><p>Instead of "In the first tutorial", I think its better to say " In the Basics of Annotation Vectors tutorial, we..."</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/TDAmeritrade/stumpy/pull/408/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

seanlaw replied:
<p>Should use <code>from stumpy import mass</code></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/TDAmeritrade/stumpy/pull/408/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

seanlaw replied:
<p>I think we can remove this section</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/TDAmeritrade/stumpy/pull/408/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

seanlaw replied:
<p>I understand the analogy but I think it's confusing to call them "stop-words"</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/TDAmeritrade/stumpy/pull/408/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='9'/>

seanlaw replied:
<p>please use snake case when defining function names</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/TDAmeritrade/stumpy/pull/408/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='11'/>

seanlaw commented on docs/Tutorial_Advanced_Annotation_Vectors.ipynb:
```diff
@@ -0,0 +1,981 @@
+{
```
<p>I can see that the EOG data is different between the two tutorials but it may not be obvious to the reader. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/TDAmeritrade/stumpy/pull/408/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='7'/>

alvii147 replied:
<p>Oh this is <strong>ECG</strong> data though, not the <strong>EOG</strong> data used in the first tutorial.</p>

seanlaw replied:
Ha! Never mind then

seanlaw commented:
> I tried reinstalling stumpy, but from stumpy import mass still doesn't work. I checked __init__.py, there was no import of mass directly from stumpy.core. I suspect it has to do with my python version, I'm using python 3.6. In any case I'm going to fix it before I push my last changes.

This is because you've likely installed STUMPY directly from `pip` or `conda`. Currently, `stumpy.mass` is only available if you install from the development version of the source code.

seanlaw commented:
As I go through the Advanced Tutorial, I feel like it may be redundant and that the first one is sufficient. I don't mind keeping the advanced one in case a user may be able to draw inspiration from it but I don't think we would publish both tutorials (on the first one and then I wouldn't even call it "basic" anymore)

alvii147 commented:
@seanlaw made some more changes according to your feedback:

- Dodgers Loop example removed
- Various grammatical improvements
- New function get_corrected_matrix_profile() defined
- Corrected Matrix Profile and Matrix Profile plots were changed to
colors mediumvioletred and darkorange respectively, so that when both
are plotted, Corrected Matrix Profile stands out more.
- "getMotifData()" replaced with "get_motif_data()"
- "window_size" replaced with "m"
- "from stumpy.core import mass" replaced with "from stumpy import mass"

You mentioned earlier that the Advanced Annotation Vectors tutorial may be unnecessary, and that the first tutorial should be enough. If that's what we want to do, give me a confirmation and I can drop the second tutorial.

seanlaw commented:
> You mentioned earlier that the Advanced Annotation Vectors tutorial may be unnecessary, and that the first tutorial should be enough. If that's what we want to do, give me a confirmation and I can drop the second tutorial.

I wouldn't drop it the second tutorial as there is value there but I would choose to leave it out of the tutorial section of our documentation. It is fine to keep it in the `docs` directory in case somebody else wanted to see more examples

alvii147 commented:
Okay that makes sense.

Do you want me to reference the second tutorial in the first tutorial then? As in, in the first tutorial it says "this is the first of a two-part tutorial series..." or "in the second part we will learn...."? Or should I change the first tutorial to sound like the second one does not exist?

seanlaw commented:
@alvii147 I think things are looking good and we are moving toward an MVP. I know this because I am starting to condense the great work that you've done down into the essential parts (i.e., "less is more"). Please be patient with me and my feedback. Your writing is excellent!

seanlaw commented:
> Do you want me to reference the second tutorial in the first tutorial then? As in, in the first tutorial it says "this is the first of a two-part tutorial series..." or "in the second part we will learn...."? Or should I change the first tutorial to sound like the second one does not exist?

Yes, please remove all references to the second tutorial from the first tutorial and simply rename the "Basics of Annotation Vectors" to "Annotation Vectors".

alvii147 commented:
> @alvii147 I think things are looking good and we are moving toward an MVP. I know this because I am starting to condense the great work that you've done down into the essential parts (i.e., "less is more"). Please be patient with me and my feedback. Your writing is excellent!

@seanlaw thanks for your detailed feedback, glad you think we are moving in the right direction. I'm gonna make more changes according to your feedback and add another set of commits. I'll remove all references of the second tutorial from the first, and rename it. I'll keep you posted.

seanlaw commented:
> I tried reinstalling stumpy, but from stumpy import mass still doesn't work. I checked __init__.py, there was no import of mass directly from stumpy.core. I suspect it has to do with my python version, I'm using python 3.6. In any case I'm going to fix it before I push my last changes.

@alvii147 Were you ultimately able to resolve this?

alvii147 commented:
> @alvii147 Were you ultimately able to resolve this?

Yepp, I solved it by installing stumpy by cloning it. In my last commit I believe I replaced `from stumpy.core import mass` to `from stumpy import mass`.

alvii147 commented:
Changes in Tutorial_Basics_Of_Annotation_Vectors.ipynb:

- Removed references to second tutorial
- Removed "Actionability Bias"
- Renamed dataframe column headers
- Used dataframe instead of the variable
- Replaced "fNIRS data" with "Hemoglobin Concentration"
- Replaced "EOG data" with "Electric Potential"
- Visualized Hemoglobin concentration data first
- Replaced motif1, motif2 with motif, neighbor respectively

Changes in Tutorial_Advanced_Annotation_Vectors.ipynb:

- Renamed dataframe column headers
- Used dataframe instead of the variable
- Replaced "ECG Data" with "Electric Potential"
- Replaced motif1, motif2 with motif, neighbor respectively

alvii147 commented:
@seanlaw I also should probably rename `Tutorial_Basics_Of_Annotation_Vectors.ipynb` to `Tutorial_Annotation_Vectors.ipynb`, since the second tutorial will not be in the read the docs page, right?

seanlaw commented:
> @seanlaw I also should probably rename `Tutorial_Basics_Of_Annotation_Vectors.ipynb` to `Tutorial_Annotation_Vectors.ipynb`, since the second tutorial will not be in the read the docs page, right?

Yes, please rename it to `Tutorial_Annotation_Vectors.ipynb`

seanlaw commented:
@alvii147 I've added a few more comments but I think we're very close. There are some minor things that I can change after the PR is merged

alvii147 commented:
@seanlaw sounds good, I've pushed another set of commits with improvements based on your feedback. Not too many changes this time.

>There are some minor things that I can change after the PR is merged
When you're merging this, look out for the `TO-DO` comments in the notebook. Some things, like the unnecessary pandas data wrangling, can be removed once the datasets are uploaded to Zenodo.

seanlaw commented:
Thank you. I will take a look

alvii147 commented:
Datasets used in `Tutorial_Annotation_Vectors.ipynb` have been replaced with Zenodo dataset links. `gdown` is no longer referenced in `Tutorial_Annotation_Vectors.ipynb`, but still used in `Tutorial_Advanced_Annotation_Vectors.ipynb`.

seanlaw commented:
> Datasets used in `Tutorial_Annotation_Vectors.ipynb` have been replaced with Zenodo dataset links. `gdown` is no longer referenced in `Tutorial_Annotation_Vectors.ipynb`, but still used in `Tutorial_Advanced_Annotation_Vectors.ipynb`.

Oops! I should've uploaded the advanced datasets too. Let me do that

seanlaw commented on docs/Tutorial_Advanced_Annotation_Vectors.ipynb:
```diff
@@ -0,0 +1,985 @@
+{
```
<p>I've uploaded the data to zenodo: https://zenodo.org/record/5067773/files/electrocardiography.csv?download=1</p><p> </p><p>Note that the column name name has been changed to "Electric Potential" (instead of "Electrical") </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/TDAmeritrade/stumpy/pull/408/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='6'/>

seanlaw replied:
<p>I have uploaded the file to Zenodo: https://zenodo.org/record/5067801/files/finger_flexion.csv?download=1</p><p> </p><p>Note that the column name has been changed to "Finger Flexion" (instead of "Finger Flexion Data")</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/TDAmeritrade/stumpy/pull/408/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='25'/>

alvii147 commented:
Updated advanced AV tut to download from Zenodo. `Tutorial_Advanced_Annotation_Vectors.ipynb` was updated to download directly from Zenodo datasets. "Electrical Potential" and "Finger Flexion Data" has been replaced with "Electric Potential" and "Finger Flexion" respectively

seanlaw commented:
> <p>@seanlaw I'm not too sure what I should replace or change in this sentence. Let me know how you want me to phrase it.</p>
> 
> ---
> 
>  <a href='https://app.reviewnb.com/TDAmeritrade/stumpy/pull/408/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

Probably not a big deal. I can probably go over this and adjust things myself.

seanlaw commented:
Thanks @alvii147 for the wonderful contribution! I appreciate all of the work that went into this.

alvii147 commented:
@seanlaw happy to help! Hopefully more contributions to come!
